Jetpack focus: remove tab bar and dashboard

AjeshRPai opened the pull request:
Part of  #17339

## Descriptions
This PR partially removes the jetpack features for users in P4. It removes the tab bar and the visible entry points for Jetpack features. This PR also disables Quick Start and the onboarding prompt in the UI in Phase 4. 

## What are the changes 
- [x] Remove the tab bar, dashboard, and initial screen setting in Jetpack feature removal phase 4 
    - [x]  Remove features
      - [x]  Stats
      - [x]  Reader
      - [x]  Notifications
    - [x] Remove Menu Items 
      - [x]  Activity Log
      - [x]  Jetpack Social
      - [x]  Backup
      - [x]  Scan
      - [x] Themes 
      - [x] People
     - [x]  Disable Quick Start


## UI 
|  Landing page in Normal Phase | Landing page in Jetpack feature removal phase 4 | 
|---|---|
| ![landing_page_light](https://user-images.githubusercontent.com/17463767/212674510-15e415ca-5ba5-4f40-8ce4-4a9abcfd6a38.png)|![landing_page_phase4_light](https://user-images.githubusercontent.com/17463767/212674519-b34e3701-1945-4d49-84e5-dee820593c19.png)| 


### Testing Instructions
- Run the app
- Open the debug menu and disable all the Jetpack Features Removal flags
- Kill the app
- Relaunch the app
- Make sure the app is behaving normally including quick start 
- Open the debug menu and enable "jp_removal_four"
- Kill the app
- Relaunch the app
- Verify that the tap bar is gone and all Jetpack features are removed.


## Regression Notes
1. Potential unintended areas of impact
App not behaving properly 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing 

3. What automated tests I added (or what prevented me from doing so)


PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

AjeshRPai committed:
Updates: the logic to determine whether the dashboard needs to be shown

AjeshRPai committed:
Updates: site items builder logic to hide site items in Phase 4

AjeshRPai committed:
Removes: Jetpack dependant configuration items on Phase 4

AjeshRPai committed:
Merge branch 'trunk' into issue/17339-remove-tab-bar-and-dashboard

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt
#	WordPress/src/main/java/org/wordpress/android/ui/mysite/items/SiteItemsBuilder.kt
#	WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt
#	WordPress/src/test/java/org/wordpress/android/ui/mysite/items/SiteItemsBuilderTest.kt

AjeshRPai committed:
Fixes: test failures and lint error

ovitrif commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt:
```diff
@@ -481,7 +486,8 @@ class MySiteViewModelTest : BaseUnitTest() {
             jetpackFeatureCardShownTracker,
-            jetpackFeatureCardHelper
+            jetpackFeatureRemovalOverlayUtil,
+                jetpackFeatureCardHelper
```
np: indentation could be fixed 👍

AjeshRPai replied:
Done in 2db424a

ovitrif replied:
Solved, Thank you 🙇

ovitrif commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/items/SiteItemsBuilderTest.kt:
```diff
@@ -34,6 +35,7 @@ class SiteItemsBuilderTest {
 
     @Mock
     lateinit var quickStartType: QuickStartType
+    @Mock lateinit var jetpackFeatureRemovalOverlayUtil: JetpackFeatureRemovalOverlayUtil
```
np: this new line-of-code could be adapted with newline after `@Mock` and a whitespace before and after it, to align with the updated code style in the test class 👍 

Same on line `50` we should align the new param with the other ones 🙏

AjeshRPai replied:
Hey ovi, 👋🏼 

Thanks for noticing it. I have updated the code to align the new param with other ones in [2db424a](https://github.com/wordpress-mobile/WordPress-Android/pull/17728/commits/2db424ae44f451ed026995a9f04e95e1ff97a712)

ovitrif replied:
Looks good, thank you for addressing this, adding to the code consistency 🙇

ovitrif reviewed (commented):
Thank for your work on this 🙇 

The majority of the changes work as described 👍 , but I noticed the following issues:

### WordPress app:
- disable all the Jetpack Features Removal flags:
  - **Expectation:** the bottom navbar is visible and I can go to: Reader, My Site or Notifications.
  - 🔴 I noticed in the WordPress app the bottom tab bar is already removed.
- enable `jp_removal_four` flag:
  - ✅ all features are removed as described
  - 🔴 If I logout and login again, after selecting a site I see this screen..
    I **expected it to be gone** since quick start is removed, and tapping `Show me around` doesn't do anything different:
<details>
   <summary>📸 Quick Start prompt in phase 4</summary>
   
   | <img width="200" src="https://user-images.githubusercontent.com/4588074/212762834-c69b7b51-f9d7-431e-82fc-efbc361a0fd7.png"> |
   | - |
 </details>  

### Jetpack app:
- Uninstall Jetpack app and install it from this branch:
  - **Expectation:** the bottom navbar is visible.
  - **🔴 Actual:** In the Jetpack app as well the bottom navbar is removed.
   My understanding is that this is not the expected behavior.

<details>
<summary>📸 Bottom navbar missing in Jetpack app</summary>

   | <img width="200" src="https://user-images.githubusercontent.com/4588074/212763580-8f1df289-aa67-4eca-9328-4a3bcb11a01e.png"> |
   | - |
</details>

  - enable `jp_removal_four` flag:
    - ✅ No unexpected feature removed

P.S. I also added two small remarks to further improve the code styling in the new unit tests code 👍 , apart from this everything else in the code LGTM 🥇 

P.P.S. Please correct me if my expectations are wrong 🙇 .

AjeshRPai committed:
Fixes: Bottom bar being shown in Jetpack app

AjeshRPai committed:
Fixes: Indent issues and formattiong

AjeshRPai committed:
Fixes: quick start being shown in Phase 4

AjeshRPai committed:
Updates: jetpack feature removal on phase 4

AjeshRPai committed:
Updates: the navigation view being shown when the flag is on or off

AjeshRPai committed:
Removes: Incorrect backstack clearing logic in phase 4

AjeshRPai commented:
Hey Ovi 👋🏼 

Thanks for the thorough review. I have fixed 
- quick start being shown in Phase 4 when logging in - [04805a0](https://github.com/wordpress-mobile/WordPress-Android/pull/17728/commits/04805a0c994d375c5aba506adbbd06f694740587) 
- bottom nav is not being shown in the Jetpack -[054f765](https://github.com/wordpress-mobile/WordPress-Android/pull/17728/commits/054f76531993178fccacf06aa17aa997fa75e837)

Although I have fixed all the issues, I am not able to fix one scenario.
- Turn on jp removal flag - Verify that the bottom nav is removed 🟢 :
- Turn off jp removal flag after turning it on - Verify that the UI is as expected 🔴 

I am having some issues trying to clear the back from the fragment manager, and the App behaves strangely. As discussed, since this is not a real word scenario plus we won't be turning off the flag once it's been on. I will take this up as a separate issue and mark it as a low-priority one. 

Video capture of the scenario to keep track of and for future reference. 

https://user-images.githubusercontent.com/17463767/212975577-691afb7a-81cb-4e16-a9a2-046928bf5ff1.mp4


With that this PR is ready for another review.

ovitrif reviewed (changes requested):
Great work fixing the identified issues so quickly, @ajesh 🥇 !

I've tested the following scenarios:

1. ✅ **Jetpack app**
   1. ✅ Verified the bottom bar is now visible
   2. ✅ Login → set `jp_removal_four`→ verify Quick Start works
2. ✅ **WordPress app**:
   1. ✅ Verified Quick Start is **not shown** in Phase 4 when logging in
   2. ✅ Verified the bottom navbar is visible
   3. `jp_removal_four`
      1. ✅ Verified the bottom navbar is hidden, the tab bar is gone and all Jetpack features are removed
      2. 🔴 Verify Quick Start dialog is hidden after login → I could not verify this because I got blocked by the crash explained below:

#### 💣 Crash when navigating the the `My Site` screen from the `Site Picker`

**Prerequisites:** (WordPress app with `jp_removal_four` enabled and logged out)

1. Open WordPress app
2. Go to device settings & enable `jp_removal_four`
3. Restart the app
4. Logout

**Repro steps:**

1. Login with WordPress.com
2. On the Site Picker Screen, select the first site
3. 💣 **Notice** the app crashes

P.S. Actually I could reproduce this bug also by force-closing the app after logout, which makes me think it's always reproducible when the `jp_removal_four`is enabled. The crash can also be reproduced by going to the Site Picker screen while logged out, and tapping the `x`button or the device back button to return to the My Site screen.

📜 **Crash Log:**

```js
FATAL EXCEPTION: main
  Process: org.wordpress.android.beta, PID: 19665
  java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=1302, result=-1, data=Intent { (has extras) }} to activity {org.wordpress.android.beta/org.wordpress.android.ui.main.WPMainActivity}: kotlin.UninitializedPropertyAccessException: lateinit property navAdapter has not been initialized
    ...
  Caused by: kotlin.UninitializedPropertyAccessException: lateinit property navAdapter has not been initialized
    at org.wordpress.android.ui.main.WPMainNavigationView.getFragment(WPMainNavigationView.kt:262)
    at org.wordpress.android.ui.main.WPMainActivity.passOnActivityResultToMySiteFragment(WPMainActivity.java:1385)
    at org.wordpress.android.ui.main.WPMainActivity.onActivityResult(WPMainActivity.java:1314)
    at android.app.Activity.dispatchActivityResult(Activity.java:8628)
    ...
```

P.P.S. I hope the shared information is useful to debug and find a fix for the crash. Not sure what's your thoughts on this, but imho this is a critical issue that should be fixed before the time `jp_removal_four` is enabled remotely.

AjeshRPai committed:
Fixes: Crash on logging in Jetpack phase 4

AjeshRPai commented:
@ovitrif .Thanks for giving me all the details about the crash. I was able to reproduce the crash.  I have fixed the crash with the commit [e3c809e](https://github.com/wordpress-mobile/WordPress-Android/pull/17728/commits/e3c809ed57d3403ed18bc0dc4f2c5c73f076072c). PR is ready for another review.

ovitrif commented:
> @ovitrif .Thanks for giving me all the details about the crash. I was able to reproduce the crash. I have fixed the crash with the commit [e3c809e](https://github.com/wordpress-mobile/WordPress-Android/pull/17728/commits/e3c809ed57d3403ed18bc0dc4f2c5c73f076072c). PR is ready for another review.

Thank you for being so prompt to address the issue with the fix @AjeshRPai 🙇 

I've validated on the latest version of this PR that [the crash reported here](https://github.com/wordpress-mobile/WordPress-Android/pull/17728#pullrequestreview-1252214882) is fixed 🚀 🎉

ovitrif reviewed (approved):
Thanks for implementing the changes and addressing all reported issues so quickly @AjeshRPai 🙇 

✅ Everything looks valid to me, both the code and the functionality with the latest changes 🎉
